### DISCUSSION TITLE
Different nodeSelector for different system components

### DISCUSSION THREAD

**Discussion by jjhidalgar** [longhorn/longhorn#9772] [General]
Hello, I'm new to Longhorn and I'm having issues differentiating the role of each component.

My assumption is that there are three types of components, at least that would make sense:
- System components (some pods that just need to run, could run anywhere)
- Disk components (the ones that take care of pysical disks)
- Storage client components (the ones needed to use the storage: csi plugin/driver)

However, it seems Longhorns makes only this division when it comes to setting nodeSelectors:
- User-deploy components
- System-managed components (also split wether they have disk or not "node.longhorn.io/create-default-disk=true")

The problem is that some components, **as my understanding** are "system components" but are longhorn managed, like:
csi-attacher csi-provisioner csi-resizer csi-snapshotter

This means that I cannot choose a different nodeSelector for these. If they are deployments (they could run anywhere) I don't want to run those on my nodes that only need to **use** storage, not provide storage, why would I.

In summary, I want to have "storage client nodes", which only run the longhorn components that are needed to physically mount the volumes, which I assume it's the longhorn-csi-plugin (and maybe engine-image-ei but not sure what that is).

----

Does it make sense what I ask?

<img width="786" alt="image" src="https://github.com/user-attachments/assets/4e0486dd-7b20-4bb4-a62f-227ac06ef465">

Ideally, storage=true would be for nodes that just **need to use storage**, but the rest of the system components that can technically run anywhere like csi-attacher (which I assume so, because it runs on nodes without create-default-disk=true, AND it is a deployment, not a daemonset) to run on infrastructure nodes. 

**But I can't set different nodeSelector on csi- and longhorn-csi-plugin, because both are set by the systemManagedComponentsNodeSelector setting.**

**Comment by akashsemil** [comment]
I do agree with the above, Here is my use case

I have a stretched Kubernetes cluster, Consider Zone A and Zone B both zones are on restricted network and will not be able to communicate to each other. I want the CSI components to only run on controlplane node so they can communicate with both the zone.

**Comment by akashsemil** [comment]
@c3y1huang - Any idea on the above ?

**Comment by mlacko64** [comment]
Hello,
I have same issue (Longhorn 1.9), I cannot use `systemManagedComponentsNodeSelector` because its settings does also apply to these two daemonsets:
```
daemonset.apps/engine-image-ei-e09d8caa   3         3         3       3            3           node.longhorn.io/create-default-disk=config   22m
daemonset.apps/longhorn-csi-plugin        3         3         3       3            3           node.longhorn.io/create-default-disk=config   21m
```

Which means if I set `systemManagedComponentsNodeSelector` on my dedicated storage nodes these two daemonsets will run just there. Also `instance-manager` pods are not able to run on other than storage nodes then, ending in pending status.

As a workaround I installed helm chart without `systemManagedComponentsNodeSelector` and then manually patched these deployments with nodeSelector pointing on my storage nodes:
```
deployment.apps/csi-attacher               3/3     3            3           23m
deployment.apps/csi-provisioner            3/3     3            3           23m
deployment.apps/csi-resizer                3/3     3            3           23m
deployment.apps/csi-snapshotter            3/3     3            3           23m
```

I believe `systemManagedComponentsNodeSelector` should not be used for both these deployments and daemonsets. Better would be have two independent parameters for helm chart (deployments vs daemonsets).
